autotraining.lua

```diff
@@ -0,0 +1,290 @@
+            if ( not state.ignored[unit.id] ) then
+                if noblePos ~=nil then
+                    for _, position in ipairs(noblePos) do
+                        if state.ignored_nobles[position.position.code] then
+                            isIgnNoble = true
+                            break
+                        end
+                    end
+                end
+                if not isIgnNoble then
+                    table.insert(ret, unit)
+                else
+                    removeTraining(unit)
+                    ignore_count = ignore_count +1
+                end
+            else
+                removeTraining(unit)
+                ignore_count = ignore_count +1
+            end
+        end
+    end
+    return ret
+end
+
+function getTrainingSquads()
+    local squads = {}
+    for squad_id, _ in pairs(state.training_squads) do
+        local squad = df.squad.find(squad_id)
+        if squad then
+            table.insert(squads, squad)
+        else
+            -- setting to nil during iteration is permitted by lua
+            state.training_squads[squad_id] = nil
+        end
+    end
+    return squads
+end
+
+function findNeed(unit)
+    local needs =  unit.status.current_soul.personality.needs
+    for _, need in ipairs(needs) do
+        if need.id == need_id then
+            return need
+        end
+    end
+    return nil
+end
+
+--######
+--Main
+--######
+
+function getByID(id)
+    for _, unit in ipairs(getTrainingCandidates()) do
+        if (unit.hist_figure_id == id) then
+            return unit
+        end
+    end
+
+    return nil
+end
+
+-- Find all training squads
+-- Abort if no squads found
+function checkSquads()
+    local squads = {}
+    for _, squad in ipairs(getTrainingSquads()) do
+        if squad.entity_id == df.global.plotinfo.group_id then
+            local leader = squad.positions[0].occupant
+            if ( leader ~= -1) then
+                table.insert(squads,squad)
+            end
+        end
+    end
+
+    if (#squads == 0) then
+        return nil
+    end
+
+    return squads
+end
+
+function addTraining(unit)
+    if (unit.military.squad_id ~= -1) then
+        for _, squad in ipairs(getTrainingSquads()) do
+            if unit.military.squad_id == squad.id then
+                return true
+            end
+        end
+        return false
+    end
+    for _, squad in ipairs(getTrainingSquads()) do
+        for i=1,9,1   do
+            if ( squad.positions[i].occupant  == -1 ) then
+                return dfhack.military.addToSquad(unit.id,squad.id,i)
+            end
+        end
+    end
+
+    return false
+end
+
+function removeTraining(unit)
+    return dfhack.military.removeFromSquad(unit.id)
+end
+
+function removeAll()
+    if ( state.training_squads == nil) then return end
+    for _, squad in ipairs(getTrainingSquads()) do
+        for i=1,9,1 do
+            local dwarf = getByID(squad.positions[i].occupant)
```

Comment:
To convert the historical figure id in the occupant field, it should suffice to use `df.historical_figure.find` and then look up the unit_id field, because `dfhack.military.removedFromSquad` takes an id instead of a unit pointer.